contracts/cw3-flex-multisig/src/contract.rs

```diff
@@ -0,0 +1,993 @@
+const MAX_LIMIT: u32 = 30;
+const DEFAULT_LIMIT: u32 = 10;
+
+fn list_proposals(
+    deps: Deps,
+    env: Env,
+    start_after: Option<u64>,
+    limit: Option<u32>,
+) -> StdResult<ProposalListResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let start = start_after.map(Bound::exclusive_int);
+    let props: StdResult<Vec<_>> = PROPOSALS
+        .range(deps.storage, start, None, Order::Ascending)
+        .take(limit)
+        .map(|p| map_proposal(&env.block, p))
+        .collect();
+
+    Ok(ProposalListResponse { proposals: props? })
+}
+
+fn reverse_proposals(
+    deps: Deps,
+    env: Env,
+    start_before: Option<u64>,
+    limit: Option<u32>,
+) -> StdResult<ProposalListResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let end = start_before.map(Bound::exclusive_int);
+    let props: StdResult<Vec<_>> = PROPOSALS
+        .range(deps.storage, None, end, Order::Descending)
+        .take(limit)
+        .map(|p| map_proposal(&env.block, p))
+        .collect();
+
+    Ok(ProposalListResponse { proposals: props? })
+}
+
+fn map_proposal(
+    block: &BlockInfo,
+    item: StdResult<(Vec<u8>, Proposal)>,
+) -> StdResult<ProposalResponse> {
+    let (key, prop) = item?;
+    let status = prop.current_status(block);
+    Ok(ProposalResponse {
+        id: parse_id(&key)?,
+        title: prop.title,
+        description: prop.description,
+        msgs: prop.msgs,
+        expires: prop.expires,
+        status,
+    })
+}
+
+fn query_vote(deps: Deps, proposal_id: u64, voter: HumanAddr) -> StdResult<VoteResponse> {
+    let voter_raw = deps.api.canonical_address(&voter)?;
+    let prop = BALLOTS.may_load(deps.storage, (proposal_id.into(), &voter_raw))?;
+    let vote = prop.map(|b| b.vote);
+    Ok(VoteResponse { vote })
+}
+
+fn list_votes(
+    deps: Deps,
+    proposal_id: u64,
+    start_after: Option<HumanAddr>,
+    limit: Option<u32>,
+) -> StdResult<VoteListResponse> {
+    let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+    let canon = maybe_canonical(deps.api, start_after)?;
+    let start = canon.map(Bound::exclusive);
+
+    let api = &deps.api;
+    let votes: StdResult<Vec<_>> = BALLOTS
+        .prefix(proposal_id.into())
+        .range(deps.storage, start, None, Order::Ascending)
+        .take(limit)
+        .map(|item| {
+            let (key, ballot) = item?;
+            Ok(VoteInfo {
+                voter: api.human_address(&CanonicalAddr::from(key))?,
+                vote: ballot.vote,
+                weight: ballot.weight,
+            })
+        })
+        .collect();
+
+    Ok(VoteListResponse { votes: votes? })
+}
+
+fn query_voter(deps: Deps, voter: HumanAddr) -> StdResult<VoterResponse> {
+    let cfg = CONFIG.load(deps.storage)?;
+    let voter_raw = deps.api.canonical_address(&voter)?;
+    let weight = cfg
+        .group
+        .is_member(&deps.querier, &voter_raw)?
+        .unwrap_or_default();
```

Comment:
I can't help but mention here that I love Rust. This level of expressiveness and power at the same time.